src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2222,12 +2222,32 @@ public override BoundNode VisitWhileStatement(BoundWhileStatement node)
             return base.VisitWhileStatement(node);
         }
 
-        public override BoundNode VisitWithExpression(BoundWithExpression expr)
+        public override BoundNode VisitWithExpression(BoundWithExpression withExpr)
         {
-            // PROTOTYPE: This is wrong
-            SetResultType(expr,
-                expr.CloneMethod?.ReturnTypeWithAnnotations.ToTypeWithState()
-                    ?? TypeWithState.Create(expr.Type, NullableFlowState.NotNull));
+            Debug.Assert(!IsConditionalState);
+
+            var receiver = withExpr.Receiver;
+            VisitRvalue(receiver);
+            _ = CheckPossibleNullReceiver(receiver, checkNullableValueType: false);
+
+            var resultType = withExpr.CloneMethod?.ReturnTypeWithAnnotations ?? ResultType.ToTypeWithAnnotations();
+            var resultSlot = GetOrCreatePlaceholderSlot(withExpr);
+            // carry over the null state of members of 'receiver' to the result of the with-expression.
+            TrackNullableStateForAssignment(receiver, resultType, resultSlot, resultType.ToTypeWithState(), MakeSlot(receiver));
+            foreach (var expr in withExpr.InitializerExpression.Initializers)
```

Comment:
>foreach [](start = 12, length = 7)

Could we re-use `VisitObjectCreationInitializer`?

Reply:
`VisitObjectCreationInitializer` requires a modification to be used. It essentially assumes that nullability warnings can never happen at the top level of the initializer. I'll make another pass to see if I can consolidate things.